src/compiler/checker.ts

```diff
@@ -393,9 +398,28 @@ namespace ts {
             }
 
             let sourceFiles = host.getSourceFiles();
-            return sourceFiles.indexOf(file1) <= sourceFiles.indexOf(file2);
-        }
+            return sourceFiles.indexOf(file1) <= sourceFiles.indexOf(file2) || !isInSameLexicalScope(); 
+
+            function isInSameLexicalScope() {
+                let node1Scope = getEnclosingBlockScopeContainer(node1);
+                let node2Scope = getEnclosingBlockScopeContainer(node2);
+                
+                while (node2Scope && node1Scope !== node2Scope) {
+                    switch (node2Scope.kind) {
+                        // In function like scope
+                        case SyntaxKind.FunctionExpression:
+                        case SyntaxKind.FunctionDeclaration:
+                        case SyntaxKind.ArrowFunction:
+                            return false;
+                    }
 
+                    node2Scope = getEnclosingBlockScopeContainer(node2Scope);

```

Comment:
Will this work if node1 is in an inner scope relative to node2? It seems like here
```ts
function foo() {
    var v1;
    function bar() {
        var v2;
    }
}
```
If you pass v1 as node1 and v2 as node2, it will return true, but if you swap the order of the arguments, it will return false.